docs/WASI-design-principles.md

```diff
@@ -0,0 +1,95 @@
+
+This has wide-ranging implications. Much of POSIX is oriented around passing around strings, whether through command-line arguments, environment variables, or paths embedded in files, with the assumption that there's a shared filesystem view between components. As we said above, we're de-emphasizing strings, which dovetails with de-emphasizing shared filesystem views. Instead of having shared state and passing around values which identify things within the shared state, WASI prefers to share as little as possible, and use handles which represent the things which need to be shared.
+
+Compatibility with existing host environments and applications is important, and we have put a lot of work into WASI libc to provide POSIX compatibility and support existing applications. There's a lot more work that can be done, and a lot more we can do to improve compatibility and user convenience. We're continuing to make progress -- and users are encouraged to [file bugs](https://github.com/WebAssembly/WASI/issues) when they find things that don't work or are awkward. This approach supports existing applications, while also supporting applications and libraries willing to opt in to enable stronger and more fine-grained security properties than are possible in regular POSIX.
+
+For example, a typical POSIX-style API might include a function that accepts a file name to open. That requires the implementation to have a filesystem view, and to have appropriate permissions within that filesystem view. WASI APIs typically prefer to instead have a function which accepts a handle for an already-open file. That way, the implementation doesn't need a filesystem view, or permissions within the filesystem. It doesn't even need to care whether there even is a filesystem. When needed, compatibility with POSIX-style APIs is then provided as a thin layer on top implementing a simple name-to-handle mapping.
+
+We recognize that this approach has tradeoffs. It often does take more work to design and implement the compatibility layers needed to support existing applications than if we just always made WASI always expose POSIX-style APIs directly. And, even when we do have compatibility mechanisms, they aren't always the most locally optimal ones. The compatibility layer overhead is usually quite modest, but it is present.
+
+However, shared-nothing linking makes it easier to use libraries in more circumstances, because you don't have to have the trust implied by a shared filesystem view, or the complexity of configuring filesystem rules for each library. With a better story for libraries and tools to work together in cooperation with the sandbox, we can build a stronger ecosystem which makes up for the downsides in the long run.
```

Comment:
"libraries" -> "programs" .. isn't shared nothing linkage really for for programs?    If you build a library with shared nothing linkage isn't it really program at that point?

Reply:
"libraries and tools" -> "programs"

Reply:
On one hand, it's a good point. Shared-nothing linking does make "libraries" look like "programs" in a sense. On the other, the core concept of a "library" is code reuse, and that's specifically what I'm looking to highlight here. If you want to reuse code, including code written by other people, using shared-nothing linking to link in that code in -- as a library -- means you don't have to worry about the implicit effects it might have on your program state.

Reply:
What is the difference between a library and tool in shared nothing linkage?

In order to make library useful in the mode wouldn't you basically need to turn it into a reactor, which is then a tool isn't it?   I know we haven't officially adopted the reactor terminology, but what would the difference be between a reactor and library that is been exposed/wrapped via SNL?

I worry that people might think they can just take a set of C libraries and build them as SNL and expect to have them work together unmodified.

Having said that I do get why library here can make sense, so I won't object to it.

Reply:
The terminology is very much in flux, but as I see it, the key thing is instance lifetimes: A command has a lifetime scoped to a call frame. A reactor lives until something (possibly itself) terminates it. And a library is something that inherits a lifetime from the context it's used in.

It's a good point that shared-nothing linking isn't going to magically fall out just by recompiling existing C code. I'll add some more words to clarify that.